src/middleware/compression.rs

```diff
@@ -0,0 +1,434 @@
+#![cfg(feature = "compression")]
+pub use accept_encoding::Encoding;
+use async_compression::stream;
+use futures::future::BoxFuture;
+use http::{header::CONTENT_ENCODING, status::StatusCode, HeaderMap};
+use http_service::{Body, Request};
+
+use crate::{
+    middleware::{Middleware, Next},
+    response::IntoResponse,
+    Context, Error, Response,
+};
+
+/// Encode settings for the compression middleware.
+///
+/// This can be modified in the case that you want more control over the speed or quality of compression.
+///
+/// For more information on how to configure each of these settings, see the async-compression crate.
+#[derive(Debug)]
+pub struct EncodeSettings {
+    pub gzip: async_compression::flate2::Compression,
+    pub deflate: async_compression::flate2::Compression,
+    pub brotli: u32,
+    pub zstd: i32,
+}
+
+impl Default for EncodeSettings {
+    fn default() -> Self {
+        Self {
+            gzip: Default::default(),
+            deflate: Default::default(),
+            brotli: 11,
+            zstd: 3,
+        }
+    }
+}
+
+/// Middleware for automatically handling outgoing response compression.
+///
+/// This middleware currently supports HTTP compression using `gzip`, `deflate`, `br`, and `zstd`.
+#[derive(Debug)]
+pub struct Compression {
+    default_encoding: Encoding,
+    settings: EncodeSettings,
+}
+
+impl Default for Compression {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Compression {
+    /// Creates a new Compression middleware. The default encoding is [`Encoding::Identity`] (no encoding).
+    pub fn new() -> Self {
+        Self {
+            default_encoding: Encoding::Identity,
+            settings: Default::default(),
+        }
+    }
+
+    /// Creates a new Compression middleware with a provided default encoding.
+    ///
+    /// This encoding will be selected if the client has not set the `Accept-Encoding` header or `*` is set as the most preferred encoding.
+    pub fn with_default(default_encoding: Encoding) -> Self {
+        Self {
+            default_encoding,
+            settings: Default::default(),
+        }
+    }
+
+    /// Accesses a mutable handle to this middleware's [`EncodeSettings`].
+    ///
+    /// This will allow you to configure this middleware's settings.
+    pub fn settings_mut(&mut self) -> &mut EncodeSettings {
+        &mut self.settings
+    }
+
+    fn preferred_encoding(&self, headers: &HeaderMap) -> Result<Encoding, Error> {
+        let encoding = match accept_encoding::parse(headers) {
+            Ok(encoding) => encoding,
+            Err(e) => return Err(Error::from(StatusCode::BAD_REQUEST)),
+        };
+        match encoding {
+            Some(encoding) => Ok(encoding),
+            None => Ok(self.default_encoding),
+        }
```

Comment:
You can use `encoding.unwrap_or(self.default_encoding)` here.

Reply:
Ah, right. I forgot about `unwrap_or`. I'll do that.

Reply:
Or maybe `unwrap_or_default` if it implements `Default`.